### PR TITLE
Support Bearer token login using long lived sessions

### PR DESCRIPTION
**Pull request by jlhughes**

### CODE REVIEW

File: src/foam/nanos/http/AuthWebAgent.java [hunk 1]
```diff
@@ -80,8 +83,9 @@ public Session authenticate(X x) {
     // context parameters
     HttpServletRequest  req          = x.get(HttpServletRequest.class);
     HttpServletResponse resp         = x.get(HttpServletResponse.class);
-    AuthService         auth         = (AuthService) x.get("auth");
+    AuthService         auth          = (AuthService) x.get("auth");
```
Comment (TW80000) [inline src/foam/nanos/http/AuthWebAgent.java]: Unnecessary space.

File: src/foam/nanos/http/AuthWebAgent.java [hunk 2]
```diff
@@ -156,10 +163,56 @@ public Session authenticate(X x) {
                 return null;
               }
             }
+          } else if ( authType.equalsIgnoreCase("bearer") ) {
+            //
+            // Support for Bearer token
+            // wget --header="Authorization: Bearer 8b4529d8-636f-a880-d0f2-637650397a71" \
+            //     http://localhost:8080/service/memory
+            //
+           String token = st.nextToken();
```
Comment (TW80000) [inline src/foam/nanos/http/AuthWebAgent.java]: Indentation

File: src/foam/nanos/http/AuthWebAgent.java [hunk 3]
```diff
@@ -147,7 +152,9 @@ public Session authenticate(X x) {
                   password = passwd;
                 }
               } else {
-                logger.debug("Invalid authorization token.");
+                logger.debug("Invalid authorization credentials. Unable to parse username:password");
```
Comment (TW80000) [inline src/foam/nanos/http/AuthWebAgent.java]: Wouldn't "Invalid authentication credentials" be more accurate? Same for elsewhere in this file.

File: src/foam/nanos/http/AuthWebAgent.java [hunk 4]
```diff
@@ -156,10 +163,56 @@ public Session authenticate(X x) {
...
+                  if ( user == null ) {
+                    throw new AuthenticationException("User not found");
+                  }
+                  Group group = user.findGroup(x);
+                  if ( group != null && ! group.getEnabled() ) {
+                    throw new AuthenticationException("Group disabled");
+                  }
+                  session.setContext(session.getContext().put("user", user).put("group", group));
+                  session = (Session) sessionDAO.put(session);
+                  logger.debug("AuthWebAgent.authenticate token login.");
```
Comment (TW80000) [inline src/foam/nanos/http/AuthWebAgent.java]: Do we need to check this in?

**Review by TW80000 — APPROVED** [review]
Tested and it works. Code looks good to me aside from a few small things. It would be nice to include the tests in the test suite so that CI will catch them if we ever regress.

File: src/foam/nanos/http/AuthWebAgent.java (hunk 1 above)
Comment (jlhughes) [inline src/foam/nanos/http/AuthWebAgent.java]:
```suggestion
    AuthService         auth         = (AuthService) x.get("auth");
```

File: src/foam/nanos/http/AuthWebAgent.java (hunk 3 above)
Comment (jlhughes) [inline src/foam/nanos/http/AuthWebAgent.java]:
```suggestion
                logger.debug("Invalid authentication credentials. Unable to parse username:password");
```

File: src/foam/nanos/http/AuthWebAgent.java [hunk 5]
```diff
@@ -147,7 +152,9 @@ public Session authenticate(X x) {
                   password = passwd;
                 }
               } else {
-                logger.debug("Invalid authorization token.");
+                logger.debug("Invalid authorization credentials. Unable to parse username:password");
+                resp.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Invalid authorization credentials.");
```
Comment (jlhughes) [inline src/foam/nanos/http/AuthWebAgent.java]:
```suggestion
                resp.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Invalid authentication credentials.");
```

File: src/foam/nanos/http/AuthWebAgent.java [hunk 6]
```diff
@@ -82,12 +85,13 @@ public Session authenticate(X x) {
     HttpServletResponse resp         = x.get(HttpServletResponse.class);
     AuthService         auth         = (AuthService) x.get("auth");
     DAO                 sessionDAO   = (DAO) x.get("localSessionDAO");
+    DAO                 userDAO      = (DAO) x.get("localUserDAO");
 
     // query parameters
     String              email        = req.getParameter("user");
-    String              password     = req.getParameter("password");
+    String              password    = req.getParameter("password");
```
Comment (jlhughes) [inline src/foam/nanos/http/AuthWebAgent.java]:
```suggestion
    String              password     = req.getParameter("password");
```

File: src/foam/nanos/http/AuthWebAgent.java [hunk 7]
```diff
@@ -82,12 +85,13 @@ public Session authenticate(X x) {
...
     String              actAs        = req.getParameter("actAs");
-    String              authHeader   = req.getHeader("Authorization");
+    String              authHeader  = req.getHeader("Authorization");
```
Comment (jlhughes) [inline src/foam/nanos/http/AuthWebAgent.java]:
```suggestion
    String              authHeader   = req.getHeader("Authorization");
```

**Pull request merged**